### PR TITLE
Improve voice quality and fix HTTP client security issues

### PR DESCRIPTION
**Pull request by treaks0**
# Improve Voice Quality and Secure HTTP Client

## Overview
This PR improves the text-to-speech (TTS) experience and hardens the HTTP client:
- More natural and consistent TTS output
- Stricter validation of backend endpoints and safer JSON parsing
- Increased test coverage around TTS and networking

##  Voice Quality

- **Premium voice selection**
  - Automatically picks the best available voice (neural / high quality, correct locale).
  - Falls back gracefully when premium voices are not available.
- **Dynamic prosody**
  - Adjusts speech rate and pitch based on sentence length and punctuation (questions, exclamations, commas).
  - Uses configurable `SpeechStyle` with sane min/max bounds to avoid distortion.
- Minor refactors for readability (use of `when`, immutable `voices` set, removal of obsolete SDK checks).

##  HTTP Client & JSON Parsing

- **HTTPS enforcement**
  - Requires HTTPS for remote endpoints.
  - Allows HTTP only for localhost/loopback and private IPs (for emulator/local dev and debug builds).
  - Removes hardcoded `10.0.2.2` and uses `InetAddress` to detect local addresses.
- **Gson-based JSON parsing**
  - Replaces manual parsing with Gson.
  - Validates presence and type of `reply` (and optional `primary_url`), trims values, and surfaces clear errors for invalid or empty responses.
- Extracts magic strings into constants (JSON keys, headers, content types, localhost identifiers).

##  Testing & Tooling

- New and extended tests for:
  - Endpoint validation (HTTPS vs HTTP, local vs remote).
  - JSON parsing edge cases (missing keys, wrong types, empty bodies).
  - TTS voice selection and prosody behavior.
- Uses `MockWebServer` for HTTP tests.
- SonarQube config fixed (encoding, test paths, dump file).
- Adds `implementation(libs.gson)` dependency.

##  Status

- All tests passing
- Build successful
- SonarQube issues resolved (security + code smells)
- No breaking API changes

Fixes PR #130 review feedback and implements voice quality improvements

### CODE REVIEW

File: app/src/main/java/com/android/sample/llm/HttpLlmClient.kt [hunk 1]
```diff
@@ -68,42 +63,97 @@ class HttpLlmClient(
...
+      address.isLoopbackAddress
+    } catch (_: UnknownHostException) {
+      // If hostname resolution fails, it's not a valid localhost address
+      false
+    }
+  }
 }
 
-internal fun parseBotReply(body: String): BotReply {
+internal fun parseBotReply(body: String, gson: Gson): BotReply {
```
Comment (hamzammt) [inline app/src/main/java/com/android/sample/llm/HttpLlmClient.kt]: Nice change.
Using a JsonObject instead of regex makes the parsing much clearer.

File: app/src/main/java/com/android/sample/speech/TextToSpeechHelper.kt [hunk 2]
```diff
@@ -320,52 +319,69 @@ class TextToSpeechHelper(
...
   /** Applies common attributes (usage, speech rate, pitch) and primes audio effects. */
   private fun configureEngine() {
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-      textToSpeech.setAudioAttributes(audioAttributes)
-    }
-    textToSpeech.setSpeechRate(DEFAULT_SPEECH_RATE)
-    textToSpeech.setPitch(DEFAULT_PITCH)
+    textToSpeech.setAudioAttributes(audioAttributes)
+    val style = speechStyle.get()
+    textToSpeech.setSpeechRate(style.baseRate)
```
Comment (hamzammt) [inline app/src/main/java/com/android/sample/speech/TextToSpeechHelper.kt]: Nice idea to centralize the TTS defaults in a SpeechStyle.
This makes it easier to tweak the base rate/pitch and update them later if needed.

File: app/src/main/java/com/android/sample/speech/TextToSpeechHelper.kt [hunk 3]
```diff
@@ -437,10 +453,82 @@ class TextToSpeechHelper(
...
+
+  fun updateSpeechStyle(style: SpeechStyle) {
+    speechStyle.set(style)
+    textToSpeech.setSpeechRate(style.baseRate)
+    textToSpeech.setPitch(style.basePitch)
+  }
+
+  private data class Prosody(val rate: Float, val pitch: Float)
+
+  private fun computeProsody(text: String): Prosody {
```
Comment (hamzammt) [inline app/src/main/java/com/android/sample/speech/TextToSpeechHelper.kt]: The rules in computeProsody() look reasonable:
– longer sentences → slow down,
– questions / exclamations → small pitch boost,
– commas → slight slowdown.
And clamping rate/pitch at the end is a good safety net.

**Review by hamzammt — APPROVED** [review]
### **Review Summary of the code :** 

Overall this looks very solid.

The TTS changes (SpeechStyle, prosody tweaks, preferred voice selection) make the behavior clearer and the voice noticeably nicer.

On the HTTP side, enforcing HTTPS for remote endpoints and using Gson with clear errors is a big improvement over the old manual parsing.

I also tested the branch locally and everything worked as expected.
Test coverage is above our threshold and there are no remaining Sonar issues, so from my side this is good to merge.
